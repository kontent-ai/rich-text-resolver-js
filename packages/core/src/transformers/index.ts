export * from "./html-transformer/index.js";
export * from "./portable-text-transformer/index.js";
export * from "./transformer-models.js";
