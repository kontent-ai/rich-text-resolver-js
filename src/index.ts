export { toVueImageDefault } from './utils/resolution/vue.js';
export { toHTMLImageDefault } from './utils/resolution/html.js';
export { traversePortableText } from './utils/transformer-utils.js';
export { transformToJson } from "./transformers/json-transformer/json-transformer.js";
export { transformToPortableText } from "./transformers/portable-text-transformer/portable-text-transformer.js";
export { parse as browserParse } from "./parser/browser/rich-text-browser-parser.js";
export { parse as nodeParse } from "./parser/node/rich-text-node-parser.js";
