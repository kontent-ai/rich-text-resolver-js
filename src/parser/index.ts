import * as runtimeEnvironment from "browser-or-node";
import { browserParse } from "./browser/index.js";
import { nodeParse } from "./node/index.js";

export * from "./parser-models.js";

export const parseHTML = (htmlInput: string) =>
  runtimeEnvironment.isBrowser
    ? browserParse(htmlInput)
    : nodeParse(htmlInput);
