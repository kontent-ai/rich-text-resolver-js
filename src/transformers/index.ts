export * from "./json-transformer/index.js";
export * from "./portable-text-transformer/index.js";
export * from "./transformer-models.js";
