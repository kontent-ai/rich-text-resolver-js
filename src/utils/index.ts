export * from "./browser-parser-utils.js";
export * from "./common-utils.js";
export * from "./constants.js";
export * from "./transformer-utils.js";
