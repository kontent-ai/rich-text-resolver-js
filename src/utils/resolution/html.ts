import {
  PortableTextHtmlComponents,
  PortableTextMarkComponent,
  PortableTextOptions,
  PortableTextTypeComponent,
  toHTML as toHTMLDefault,
} from "@portabletext/to-html";

import {
  PortableTextComponentOrItem,
  PortableTextExternalLink,
  PortableTextImage,
  PortableTextItemLink,
  PortableTextObject,
  PortableTextTable,
  PortableTextTableCell,
  PortableTextTableRow,
} from "../../index.js";

type RichTextCustomBlocks = Partial<{
  image: PortableTextTypeComponent<PortableTextImage>;
  componentOrItem: PortableTextTypeComponent<PortableTextComponentOrItem>;
  table: PortableTextTypeComponent<PortableTextTable>;
}>;

type RichTextCustomMarks = Partial<{
  contentItemLink: PortableTextMarkComponent<PortableTextItemLink>;
  link: PortableTextMarkComponent<PortableTextExternalLink>;
}>;

type RichTextHtmlComponents = Omit<PortableTextHtmlComponents, "types" | "marks"> & {
  types: PortableTextHtmlComponents["types"] & RichTextCustomBlocks;
  marks: PortableTextHtmlComponents["marks"] & RichTextCustomMarks;
};

/**
 * Converts array of portable text objects to HTML. Optionally, custom resolvers can be provided.
 *
 * This function is a wrapper around `toHTML` function from `@portabletext/to-html` package, with default resolvers for `sup` and `sub` marks added.
 *
 * @param blocks array of portable text objects
 * @param resolvers optional custom resolvers for Portable Text objects
 * @returns HTML string
 */
export const toHTML = (blocks: PortableTextObject[], resolvers?: PortableTextHtmlResolvers) => {
  const defaultComponentResolvers: PortableTextHtmlResolvers = {
    components: {
      types: {
        image: ({ value }) => resolveImage(value),
        table: ({ value }) => resolveTable(value),
      },
      marks: {
        link: ({ value, children }) => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { _key, _type, ...attributes } = value!;
          return `<a ${
            Object.entries(attributes)
              .map(([key, value]) => `${key}="${value}"`)
              .join(" ")
          }>${children}</a>`;
        },
        sup: ({ children }) => `<sup>${children}</sup>`,
        sub: ({ children }) => `<sub>${children}</sub>`,
      },
    },
  };

  const mergedComponentResolvers = {
    types: {
      ...defaultComponentResolvers.components.types,
      ...resolvers?.components.types,
    },
    marks: {
      ...defaultComponentResolvers.components.marks,
      ...resolvers?.components.marks,
    },
  };

  return toHTMLDefault(blocks, { components: mergedComponentResolvers });
};

/**
 * Extends `PortableTextOptions` type from `toHTML` package with
 * pre-defined types for resolution of Kontent.ai specific custom objects.
 */
export type PortableTextHtmlResolvers = Omit<PortableTextOptions, "components"> & {
  components: Partial<RichTextHtmlComponents>;
};

/**
 * Renders a portable text table to HTML.
 *
 * @param {PortableTextTable} table - The portable text table object to render.
 * @param {(value: PortableTextObject[]) => string} resolver - A function that resolves
 *        the content of each cell in the table.
 * @returns {string} The rendered table as an HTML string.
 */
export const resolveTable = (
  table: PortableTextTable,
  resolver: (value: PortableTextObject[]) => string = toHTML,
) => {
  const renderCell = (cell: PortableTextTableCell) => {
    const cellContent = resolver(cell.content);
    return `<td>${cellContent}</td>`;
  };

  const renderRow = (row: PortableTextTableRow) => {
    const cells = row.cells.map(renderCell);
    return `<tr>${cells.join("")}</tr>`;
  };

  const renderRows = () => table.rows.map(renderRow).join("");

  return `<table><tbody>${renderRows()}</tbody></table>`;
};

/**
 * Resolves an image object to HTML.
 *
 * @param {PortableTextImage} image - The portable text image object to be rendered.
 * @param {(image: PortableTextImage) => string} resolver - A resolver function that returns the image as an HTML string. Default implementation provided if not specified.
 * @returns {string} The resolved image as an HTML string.
 */
export const resolveImage = (
  image: PortableTextImage,
  resolver: (image: PortableTextImage) => string = toHTMLImageDefault,
): string => resolver(image);

/**
 * Provides a default resolver function for an image object to HTML. This function can be used as
 * a default argument for the `resolveImage` function.
 *
 * @param {PortableTextImage} image - The portable text image object to be rendered.
 * @returns {VueImage} An object representing the image, containing `src` and `alt` properties,
 *          and potentially other HTML attributes.
 */
export const toHTMLImageDefault = (image: PortableTextImage): string =>
  `<img src="${image.asset.url}" alt="${image.asset.alt}">`;
