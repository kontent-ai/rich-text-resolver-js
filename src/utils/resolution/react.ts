import { PortableTextMarkComponent, PortableTextReactComponents, PortableTextTypeComponent } from "@portabletext/react";

import {
  PortableTextComponentOrItem,
  PortableTextExternalLink,
  PortableTextImage,
  PortableTextItemLink,
  PortableTextTable,
} from "../../transformers/transformer-models.js";

type RichTextCustomBlocks = {
  image: PortableTextTypeComponent<PortableTextImage>;
  componentOrItem: PortableTextTypeComponent<PortableTextComponentOrItem>;
  table: PortableTextTypeComponent<PortableTextTable>;
};

type RichTextCustomMarks = {
  contentItemLink: PortableTextMarkComponent<PortableTextItemLink>;
  link: PortableTextMarkComponent<PortableTextExternalLink>;
};

/**
 * Extends `PortableTextReactComponents` type from `@portabletext/react` package with
 * pre-defined types for resolution of Kontent.ai specific custom objects.
 */
export type PortableTextReactResolvers = Partial<
  Omit<
    PortableTextReactComponents,
    "types" | "marks"
  > & {
    types: PortableTextReactComponents["types"] & Partial<RichTextCustomBlocks>;
    marks: PortableTextReactComponents["marks"] & Partial<RichTextCustomMarks>;
  }
>;
