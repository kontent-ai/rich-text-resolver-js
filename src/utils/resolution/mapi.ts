import {
  PortableTextMarkComponentOptions,
  PortableTextOptions,
  PortableTextTypeComponentOptions,
  toHTML,
} from "@portabletext/to-html";

import {
  PortableTextComponentOrItem,
  PortableTextExternalLink,
  PortableTextImage,
  PortableTextItemLink,
  PortableTextMark,
  PortableTextObject,
  PortableTextTable,
} from "../../transformers/transformer-models.js";
import { throwError } from "../common-utils.js";
import { resolveTable } from "./html.js";

const toManagementApiImage = (image: PortableTextImage) => createFigureTag(image.asset._ref);

const toManagementApiRichTextItem = (richTextItem: PortableTextComponentOrItem) =>
  `<object type="application/kenticocloud" data-type="${richTextItem.dataType}" data-id="${richTextItem.component._ref}"></object>`;

const toManagementApiTable = (table: PortableTextTable) =>
  resolveTable(table, (blocks) => toHTML(blocks, portableTextComponents));

const toManagementApiExternalLink = (children: string, link?: PortableTextExternalLink) =>
  link
    ? `<a ${createExternalLinkAttributes(link)}>${children}</a>`
    : throwError("Mark definition for external link not found.");

const toManagementApiItemLink = (children: string, link?: PortableTextItemLink) =>
  link
    ? `<a data-item-id="${link.reference._ref}">${children}</a>`
    : throwError("Mark definition for item link not found.");

const createImgTag = (assetId: string) => `<img src="#" data-asset-id="${assetId}">`;

const createFigureTag = (assetId: string) => `<figure data-asset-id="${assetId}">${createImgTag(assetId)}</figure>`;

const createExternalLinkAttributes = (link: PortableTextExternalLink) =>
  Object.entries(link).filter(([k]) => k !== "_type" && k !== "_key").map(([k, v]) => `${k}="${v}"`).join(" ");

/**
 * specifies resolution for custom types and marks that are not part of `toHTML` default implementation.
 */
const portableTextComponents: PortableTextOptions = {
  components: {
    types: {
      image: ({ value }: PortableTextTypeComponentOptions<PortableTextImage>) => toManagementApiImage(value),
      component: ({ value }: PortableTextTypeComponentOptions<PortableTextComponentOrItem>) =>
        toManagementApiRichTextItem(value),
      table: ({ value }: PortableTextTypeComponentOptions<PortableTextTable>) => toManagementApiTable(value),
    },
    marks: {
      contentItemLink: ({ children, value }: PortableTextMarkComponentOptions<PortableTextItemLink>) =>
        toManagementApiItemLink(children, value),
      link: ({ children, value }: PortableTextMarkComponentOptions<PortableTextExternalLink>) =>
        toManagementApiExternalLink(children, value),
      sub: ({ children }: PortableTextMarkComponentOptions<PortableTextMark>) => `<sub>${children}</sub>`,
      strong: ({ children }: PortableTextMarkComponentOptions<PortableTextMark>) => `<strong>${children}</strong>`,
      sup: ({ children }: PortableTextMarkComponentOptions<PortableTextMark>) => `<sup>${children}</sup>`,
      em: ({ children }: PortableTextMarkComponentOptions<PortableTextMark>) => `<em>${children}</em>`,
    },
  },
};

/**
 * Transforms Portable Text initially created from Kontent.ai management API rich text back to MAPI-compatible format.
 *
 * This function performs only minimal checks for compatibility and is therefore not suited for conversion of generic HTML
 * or any other rich text other than MAPI format.
 *
 * @param blocks portable text array
 * @returns MAPI-compatible rich text string
 */
export const toManagementApiFormat = (blocks: PortableTextObject[]) => toHTML(blocks, portableTextComponents);
