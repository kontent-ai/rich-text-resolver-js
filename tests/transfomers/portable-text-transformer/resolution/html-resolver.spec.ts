import { Elements, ElementType } from "@kontent-ai/delivery-sdk";
import { PortableTextTypeComponentOptions } from "@portabletext/to-html";

import { PortableTextComponentOrItem, transformToPortableText } from "../../../../src";
import { PortableTextHtmlResolvers, toHTML } from "../../../../src/utils/resolution/html";

jest.mock("short-unique-id", () => {
  return jest.fn().mockImplementation(() => {
    return {
      randomUUID: jest.fn().mockReturnValue("guid"),
    };
  });
});

describe("HTML resolution", () => {
  const richTextInput: Elements.RichTextElement = {
    value: "<p><br></p>",
    type: ElementType.RichText,
    images: [],
    linkedItemCodenames: [],
    linkedItems: [
      {
        system: {
          id: "99e17fe7-a215-400d-813a-dc3608ee0294",
          name: "test item",
          codename: "test_item",
          language: "default",
          type: "test",
          collection: "default",
          sitemapLocations: [],
          lastModified: "2022-10-11T11:27:25.4033512Z",
          workflowStep: "published",
          workflow: "default",
        },
        elements: {
          text_element: {
            type: ElementType.Text,
            name: "text element",
            value: "random text value",
          },
        },
      },
    ],
    links: [],
    name: "dummy",
  };

  const customResolvers: PortableTextHtmlResolvers = {
    components: {
      block: {
        h1: ({ children }) => `<h1 custom-attribute="value">${children}</h1>`,
      },
      types: {
        image: ({ value }) => `<img src="${value.asset.url}" alt="${value.asset.rel ?? ""}" height="800">`,
        componentOrItem: ({
          value,
        }: PortableTextTypeComponentOptions<PortableTextComponentOrItem>) => {
          const linkedItem = richTextInput.linkedItems.find(
            (item) => item.system.codename === value.componentOrItem._ref,
          );
          if (!linkedItem) {
            return `Resolver for unknown type not implemented.`;
          }

          switch (linkedItem.system.type) {
            case "test":
              return `<p>resolved value of text_element: <strong>${linkedItem.elements.text_element.value}</strong></p>`;
            default:
              return `Resolver for type ${linkedItem.system.type} not implemented.`;
          }
        },
      },
      marks: {
        contentItemLink: ({ children, value }) =>
          `<a href="https://website.com/${value?.contentItemLink._ref}">${children}</a>`,
        sup: ({ children }) => `<sup custom-attribute="value">${children}</sup>`,
      },
    },
  };

  const transformAndCompare = (
    richTextValue: string,
    customResolvers?: PortableTextHtmlResolvers,
  ) => {
    const portableText = transformToPortableText(richTextValue);
    const result = toHTML(portableText, customResolvers);

    expect(result).toMatchSnapshot();
  };

  it("builds basic portable text into HTML", () => {
    transformAndCompare(
      "<p><br></p><p>text<a href=\"http://google.com\" data-new-window=\"true\" title=\"linktitle\" target=\"_blank\" rel=\"noopener noreferrer\"><strong>link</strong></a></p><h1>heading</h1><p><br></p>",
    );
  });

  it("resolves internal link", () => {
    transformAndCompare(
      "<p><a data-item-id=\"23f71096-fa89-4f59-a3f9-970e970944ec\" href=\"\"><em>item</em></a></p>",
      customResolvers,
    );
  });

  it("resolves a linked item", () => {
    transformAndCompare(
      "<object type=\"application/kenticocloud\" data-type=\"item\" data-rel=\"link\" data-codename=\"test_item\"></object><p>text after component</p>",
      customResolvers,
    );
  });

  it("resolves a table using default fallback", () => {
    transformAndCompare(
      "<table><tbody>\n  <tr><td>Ivan</td><td>Jiri</td></tr>\n  <tr><td>Ondra</td><td>Dan</td></tr>\n</tbody></table>",
    );
  });

  it("resolves an asset", () => {
    transformAndCompare(
      "<figure data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\"><img src=\"https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg\" data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" alt=\"\"></figure>",
    );
  });

  it("resolves an asset with custom resolver", () => {
    transformAndCompare(
      "<figure data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\"><img src=\"https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg\" data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" alt=\"\"></figure>",
      customResolvers,
    );
  });

  it("resolves styled text with line breaks", () => {
    transformAndCompare(
      "</p>\n<p><strong>Strong text with line break<br>\nStrong text with line break</strong></p>",
    );
  });

  it("by default resolves text styled with sub and sup", () => {
    transformAndCompare(
      "<p><sub>Subscript text</sub><sup>Superscript text</sup></p>",
    );
  });

  it("resolves superscript with custom resolver", () => {
    transformAndCompare("<p><sup>Superscript text</sup></p>", customResolvers);
  });

  it("resolves a link using default fallback", () => {
    transformAndCompare(
      "<p><a href=\"https://website.com/12345\" target=\"_blank\" rel=\"noopener noreferrer\">link</a></p>",
    );
  });

  it("uses custom resolver for image, fallbacks to default for a table", () => {
    transformAndCompare(
      "<table><tbody>\n  <tr><td>Ivan</td><td>Jiri</td></tr>\n  <tr><td>Ondra</td><td>Dan</td></tr>\n</tbody></table><img src=\"https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg\" data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" alt=\"\">",
      customResolvers,
    );
  });

  it("resolves a heading using default fallback", () => {
    transformAndCompare("<h1>heading</h1>");
  });

  it("resolves a heading using custom resolvers", () => {
    transformAndCompare("<h1>heading</h1>", customResolvers);
  });

  it("resolves a h2 heading while using custom resolvers for h1", () => {
    transformAndCompare(
      "<h1>modified heading</h1><h2>heading</h2>",
      customResolvers,
    );
  });
});
