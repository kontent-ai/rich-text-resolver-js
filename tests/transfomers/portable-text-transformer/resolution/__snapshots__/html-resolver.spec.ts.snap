// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML resolution builds basic portable text into HTML 1`] = `"<p><br/></p><p>text<a href="http://google.com" data-new-window="true" title="linktitle" target="_blank" rel="noopener noreferrer"><strong>link</strong></a></p><h1>heading</h1><p><br/></p>"`;

exports[`HTML resolution by default resolves text styled with sub and sup 1`] = `"<p><sub>Subscript text</sub><sup>Superscript text</sup></p>"`;

exports[`HTML resolution resolves a link using default fallback 1`] = `"<p><a href="https://website.com/12345" target="_blank" rel="noopener noreferrer">link</a></p>"`;

exports[`HTML resolution resolves a linked item 1`] = `"<p>resolved value of text_element: <strong>random text value</strong></p><p>text after component</p>"`;

exports[`HTML resolution resolves a table using default fallback 1`] = `"<table><tbody><tr><td><p>Ivan</p></td><td><p>Jiri</p></td></tr><tr><td><p>Ondra</p></td><td><p>Dan</p></td></tr></tbody></table>"`;

exports[`HTML resolution resolves an asset 1`] = `"<img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg" alt="">"`;

exports[`HTML resolution resolves an asset with custom resolver 1`] = `"<img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg" alt="" height="800">"`;

exports[`HTML resolution resolves internal link 1`] = `"<p><a href="https://website.com/23f71096-fa89-4f59-a3f9-970e970944ec"><em>item</em></a></p>"`;

exports[`HTML resolution resolves styled text with line breaks 1`] = `"<p><strong>Strong text with line break<br/>Strong text with line break</strong></p>"`;

exports[`HTML resolution resolves superscript with custom resolver 1`] = `"<p><sup custom-attribute="value">Superscript text</sup></p>"`;

exports[`HTML resolution uses custom resolver for image, fallbacks to default for a table 1`] = `"<table><tbody><tr><td><p>Ivan</p></td><td><p>Jiri</p></td></tr><tr><td><p>Ondra</p></td><td><p>Dan</p></td></tr></tbody></table><img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg" alt="" height="800">"`;
